File: test/cs/Ssh.Test/ServiceTests.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.IO;
using Microsoft.DevTunnels.Ssh.Messages;
using Microsoft.DevTunnels.Ssh.Services;
using Xunit;
using Xunit.Sdk;

namespace Microsoft.DevTunnels.Ssh.Test;

public class ServiceTests : IDisposable
{
	private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(20);

	private readonly TestServiceConfig testConfig = new TestServiceConfig();
	private readonly SessionPair sessionPair;
	private readonly SshClientSession clientSession;
	private readonly SshServerSession serverSession;
	private SemaphoreSlim clientClosedSemaphore = new SemaphoreSlim(0);
	private SemaphoreSlim serverClosedSemaphore = new SemaphoreSlim(0);
	private SshSessionClosedEventArgs clientClosedEvent = null;
	private SshSessionClosedEventArgs serverClosedEvent = null;
	private SshService activatedService = null;

	public ServiceTests()
	{
		var clientConfig = new SshSessionConfiguration();
		var serverConfig = new SshSessionConfiguration();
		clientConfig.Services.Add(typeof(MessageService), null);
		serverConfig.Services.Add(typeof(TestService1), null);
		serverConfig.Services.Add(typeof(TestService2), testConfig);
		serverConfig.Services.Add(typeof(TestService3), testConfig);
		serverConfig.Services.Add(typeof(TestService4), null);
		serverConfig.Services.Add(typeof(TestService5), null);

		this.sessionPair = new SessionPair(serverConfig, clientConfig);
		this.clientSession = sessionPair.ClientSession;
		this.serverSession = sessionPair.ServerSession;

		this.serverSession.ServiceActivated += (sender, e) =>
		{
			this.activatedService = e;
		};

		this.clientSession.Closed += (sender, e) =>
		{
			this.clientClosedEvent = e;
			this.clientClosedSemaphore.Release();
		};
		this.serverSession.Closed += (sender, e) =>
		{
			this.serverClosedEvent = e;
			this.serverClosedSemaphore.Release();
		};
	}

	public void Dispose()
	{
		this.clientSession.Dispose();
		this.serverSession.Dispose();
	}

	private async Task CloseSessions()
	{
		this.sessionPair.ServerStream.Close();
		this.sessionPair.ClientStream.Close();

		await Task.WhenAll(
			this.serverClosedSemaphore.WaitAsync(),
			this.clientClosedSemaphore.WaitAsync()).WithTimeout(Timeout);
	}

	[Fact]
	public async Task ActivateOnServiceRequest()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService1);

		await this.clientSession.RequestServiceAsync(TestService1.Name).WithTimeout(Timeout);
		Assert.Contains(this.serverSession.Services, (s) => s is TestService1);
		Assert.IsType<TestService1>(this.activatedService);

		var testService = this.serverSession.Services.OfType<TestService1>().Single();
		Assert.Null(testService.Config);

		Assert.False(testService.DisposedTask.IsCompleted);
		await this.CloseSessions();
		await testService.DisposedTask.WithTimeout(Timeout);
	}

	[Fact]
	public async Task ActivateOnSessionRequest()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService2);

		var request = new SessionRequestMessage
		{
			RequestType = TestService2.Request,
			WantReply = true,
		};
		var result = await this.clientSession.RequestAsync(request).WithTimeout(Timeout);
		Assert.True(result);
		Assert.Contains(this.serverSession.Services, (s) => s is TestService2);
		Assert.IsType<TestService2>(this.activatedService);

		var testService = this.serverSession.Services.OfType<TestService2>().Single();
		Assert.Equal(testConfig, testService.Config);
		Assert.NotNull(testService.RequestMessage);

		await this.CloseSessions();
	}

	[Fact]
	public async Task ActivateOnChannelType()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService3);

		var channel = await this.clientSession.OpenChannelAsync(TestService3.ChannelType)
			.WithTimeout(Timeout);
		Assert.Contains(this.serverSession.Services, (s) => s is TestService3);
		Assert.IsType<TestService3>(this.activatedService);

		var testService = this.serverSession.Services.OfType<TestService3>().Single();
		Assert.Equal(testConfig, testService.Config);
		Assert.NotNull(testService.Channel);

		await this.CloseSessions();
	}

	[Fact]
	public async Task ActivateOnChannelRequest()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService4);

		var channel = await this.clientSession.OpenChannelAsync()
			.WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService4);

		var request = new ChannelRequestMessage
		{
			RequestType = TestService4.Request,
			WantReply = true,
		};
		await channel.RequestAsync(request).WithTimeout(Timeout);
		Assert.Contains(this.serverSession.Services, (s) => s is TestService4);
		Assert.IsType<TestService4>(this.activatedService);

		var testService = this.serverSession.Services.OfType<TestService4>().Single();
		Assert.NotNull(testService.RequestMessage);

		await this.CloseSessions();
	}

	[Fact]
	public async Task ActivateOnChannelTypeChannelRequest()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService5);

		var channel = await this.clientSession.OpenChannelAsync(TestService5.ChannelType)
			.WithTimeout(Timeout);
		Assert.DoesNotContain(this.serverSession.Services, (s) => s is TestService5);

		var request = new ChannelRequestMessage
		{
			RequestType = TestService5.Request,
			WantReply = true,
		};
		await channel.RequestAsync(request).WithTimeout(Timeout);
		Assert.Contains(this.serverSession.Services, (s) => s is TestService5);
		Assert.IsType<TestService5>(this.activatedService);

		var testService = this.serverSession.Services.OfType<TestService5>().Single();
		Assert.NotNull(testService.Channel);
		Assert.NotNull(testService.RequestMessage);

		await this.CloseSessions();
	}

	[Fact]
	public async Task SendUnimplementedMessage()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);

		var messageService = this.sessionPair.ClientSession.ActivateService<MessageService>();
		await messageService.SendMessageAsync(new TestMessage(), CancellationToken.None);

		// Wait for open channel response, to ensure the message is processed.
		await this.sessionPair.ClientSession.OpenChannelAsync();
	}

	[Fact]
	public async Task SendDebugMessage()
	{
		await this.sessionPair.ConnectAsync().WithTimeout(Timeout);

		var messageService = this.sessionPair.ClientSession.ActivateService<MessageService>();
		await messageService.SendMessageAsync(new DebugMessage("test"), CancellationToken.None);

		// Wait for open channel response, to ensure the message is processed.
		await this.sessionPair.ClientSession.OpenChannelAsync();
	}

	private class TestServiceConfig
	{
	}

	[ServiceActivation(ServiceRequest = Name)]
	private class TestService1 : SshService
	{
		public const string Name = "test-service-1";
		private readonly TaskCompletionSource<bool> disposedCompletion =
			new TaskCompletionSource<bool>();

		public TestService1(SshSession session) : base(session) { }
		public TestService1(SshSession session, TestServiceConfig config) : base(session)
		{
			Config = config;
		}

		public TestServiceConfig Config { get; }

		public Task<bool> DisposedTask => this.disposedCompletion.Task;

		protected override void Dispose(bool disposing)
		{
			this.disposedCompletion.TrySetResult(disposing);
			base.Dispose(disposing);
		}
	}

	[ServiceActivation(SessionRequest = Request)]
	private class TestService2 : SshService
	{
		public const string Request = "test-service-2";

		public TestService2(SshSession session) : base(session) { }
		public TestService2(SshSession session, TestServiceConfig config) : base(session)
		{
			Config = config;
		}

		public TestServiceConfig Config { get; }

		public SessionRequestMessage RequestMessage { get; private set; }

		protected override async Task OnSessionRequestAsync(
			SshRequestEventArgs<SessionRequestMessage> request,
			CancellationToken cancellation)
		{
			RequestMessage = request.Request;
			request.IsAuthorized = true;

			if (request.Request.WantReply)
			{
				await SendMessageAsync(new SessionRequestSuccessMessage(), cancellation);
			}
		}
	}

	[ServiceActivation(ChannelType = ChannelType)]
	private class TestService3 : SshService
	{
		public const string ChannelType = "test-service-3";

		public TestService3(SshSession session, TestServiceConfig config) : base(session)
		{
			Config = config;
		}

		public TestServiceConfig Config { get; }

		public SshChannel Channel { get; private set; }

		protected override Task<ChannelMessage> OnChannelOpeningAsync(
			SshChannelOpeningEventArgs args,
			CancellationToken cancellation)
		{
			Channel = args.Channel;
			return Task.FromResult<ChannelMessage>(new ChannelOpenConfirmationMessage());
		}
	}

	[ServiceActivation(ChannelRequest = Request)]
	private class TestService4 : SshService
	{
		public const string Request = "test-service-4";

		public TestService4(SshSession session) : base(session) { }

		public ChannelRequestMessage RequestMessage { get; private set; }

		public SshChannel Channel { get; private set; }

		protected override Task OnChannelRequestAsync(
			SshChannel channel,
			SshRequestEventArgs<ChannelRequestMessage> request,
			CancellationToken cancellation)
		{
			Channel = channel;
			RequestMessage = request.Request;
			request.IsAuthorized = true;
			return Task.CompletedTask;
		}
	}

	[ServiceActivation(ChannelType = ChannelType, ChannelRequest = Request)]
	private class TestService5 : SshService
	{
		public const string ChannelType = "test-service-5-channel";
		public const string Request = "test-service-5";

		public TestService5(SshSession session) : base(session) { }

		public SshChannel Channel { get; private set; }

		public ChannelRequestMessage RequestMessage { get; private set; }

		protected override Task OnChannelRequestAsync(
			SshChannel channel,
			SshRequestEventArgs<ChannelRequestMessage> request,
			CancellationToken cancellation)
		{
			Channel = channel;
			RequestMessage = request.Request;
			request.IsAuthorized = true;
			return Task.CompletedTask;
		}
	}

	/// <summary>
	/// A test service capable of sending custom message types.
	/// </summary>
	[ServiceActivation(ServiceRequest = Name)]
	private class MessageService : SshService
	{
		public const string Name = "test-message-service";

		public MessageService(SshSession session) : base(session) { }

		public new Task SendMessageAsync(SshMessage message, CancellationToken cancellation)
			=> base.SendMessageAsync(message, cancellation);
	}

	private class TestMessage : SshMessage
	{
		public override byte MessageType => 199;

		protected override void OnRead(ref SshDataReader reader)
		{
		}

		protected override void OnWrite(ref SshDataWriter writer)
		{
		}
	}
}

File: test/ts/ssh-test/serviceTests.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as assert from 'assert';
import { suite, test, slow, timeout, pending } from '@testdeck/mocha';

import {
	CancellationToken,
	ChannelRequestMessage,
	DebugMessage,
	KeyPair,
	PromiseCompletionSource,
	serviceActivation,
	SessionRequestMessage,
	SessionRequestSuccessMessage,
	SshAlgorithms,
	SshChannel,
	SshChannelOpeningEventArgs,
	SshClientSession,
	SshDataReader,
	SshDataWriter,
	SshDisconnectReason,
	SshRequestEventArgs,
	SshServerSession,
	SshService,
	SshSession,
	SshMessage,
} from '@microsoft/dev-tunnels-ssh';
import { connectSessionPair, createSessionConfig } from './sessionPair';

@suite
@slow(3000)
@timeout(20000)
export class ServiceTests {
	private static readonly testUsername = 'test';

	private readonly testConfig = new TestServiceConfig();
	private clientClosedCompletion = new PromiseCompletionSource<void>();
	private serverClosedCompletion = new PromiseCompletionSource<void>();
	private activatedService: SshService | null = null;

	private static serverKey: KeyPair;

	@slow(10000)
	@timeout(20000)
	public static async before() {
		ServiceTests.serverKey = await SshAlgorithms.publicKey.rsaWithSha512!.generateKeyPair();
	}

	private async createSessions(): Promise<[SshClientSession, SshServerSession]> {
		var clientConfig = createSessionConfig();
		var serverConfig = createSessionConfig();
		clientConfig.services.set(MessageService, null);
		serverConfig.services.set(TestService1, null);
		serverConfig.services.set(TestService2, this.testConfig);
		serverConfig.services.set(TestService3, this.testConfig);
		serverConfig.services.set(TestService4, null);
		serverConfig.services.set(TestService5, null);

		const clientSession = new SshClientSession(clientConfig);
		const serverSession = new SshServerSession(serverConfig);

		serverSession.credentials.publicKeys = [ServiceTests.serverKey];

		serverSession.onAuthenticating((e) => (e.authenticationPromise = Promise.resolve({})));
		clientSession.onAuthenticating((e) => (e.authenticationPromise = Promise.resolve({})));

		serverSession.onServiceActivated((service) => {
			this.activatedService = service;
		});

		clientSession.onClosed((e) => {
			this.clientClosedCompletion.resolve();
		});
		serverSession.onClosed((e) => {
			this.serverClosedCompletion.resolve();
		});

		return [clientSession, serverSession];
	}

	private async closeSessions(
		clientSession: SshSession,
		serverSession: SshSession,
	): Promise<void> {
		await clientSession.close(SshDisconnectReason.byApplication, 'test');
		await serverSession.close(SshDisconnectReason.byApplication, 'test');

		await Promise.all([this.serverClosedCompletion.promise, this.clientClosedCompletion.promise]);
	}

	@test
	public async activateOnServiceRequest(): Promise<void> {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		assert(!serverSession.services.find((s) => s instanceof TestService1));

		await clientSession.requestService(TestService1.serviceName);
		assert(serverSession.services.find((s) => s instanceof TestService1));
		assert(this.activatedService instanceof TestService1);

		const testService = <TestService1>(
			serverSession.services.find((s) => s instanceof TestService1)
		);
		assert.equal(testService.config, null);

		assert(!testService.isDisposed);
		await this.closeSessions(clientSession, serverSession);
		await testService.disposedPromise;
	}

	@test
	public async activateOnSessionRequest(): Promise<void> {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		assert(!serverSession.services.find((s) => s instanceof TestService2));

		const request = new SessionRequestMessage(TestService2.requestName, true);

		var result = await clientSession.request(request);
		assert(result);
		assert(serverSession.services.find((s) => s instanceof TestService2));
		assert(this.activatedService instanceof TestService2);

		const testService = <TestService2>(
			serverSession.services.find((s) => s instanceof TestService2)
		);
		assert.equal(this.testConfig, testService.config);
		assert(testService.requestMessage);

		await this.closeSessions(clientSession, serverSession);
	}

	@test
	public async activateOnChannelType(): Promise<void> {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		assert(!serverSession.services.find((s) => s instanceof TestService3));

		await clientSession.openChannel(TestService3.channelType);
		assert(serverSession.services.find((s) => s instanceof TestService3));
		assert(this.activatedService instanceof TestService3);

		const testService = <TestService3>(
			serverSession.services.find((s) => s instanceof TestService3)
		);
		assert.equal(this.testConfig, testService.config);
		assert(testService.channel);

		await this.closeSessions(clientSession, serverSession);
	}

	@test
	public async activateOnChannelRequest(): Promise<void> {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		assert(!serverSession.services.find((s) => s instanceof TestService4));

		const channel = await clientSession.openChannel();
		assert(!serverSession.services.find((s) => s instanceof TestService4));

		const request = new ChannelRequestMessage();
		request.requestType = TestService4.requestName;
		request.wantReply = true;
		await channel.request(request);
		assert(serverSession.services.find((s) => s instanceof TestService4));
		assert(this.activatedService instanceof TestService4);

		const testService = <TestService4>(
			serverSession.services.find((s) => s instanceof TestService4)
		);
		assert(testService.requestMessage);

		await this.closeSessions(clientSession, serverSession);
	}

	@test
	public async activateOnChannelTypeChannelRequest(): Promise<void> {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		assert(!serverSession.services.find((s) => s instanceof TestService5));

		var channel = await clientSession.openChannel(TestService5.channelType);
		assert(!serverSession.services.find((s) => s instanceof TestService5));

		const request = new ChannelRequestMessage();
		request.requestType = TestService5.requestName;
		request.wantReply = true;
		await channel.request(request);
		assert(serverSession.services.find((s) => s instanceof TestService5));
		assert(this.activatedService instanceof TestService5);

		const testService = <TestService5>(
			serverSession.services.find((s) => s instanceof TestService5)
		);
		assert(testService.channel);
		assert(testService.requestMessage);

		await this.closeSessions(clientSession, serverSession);
	}

	@test
	public async sendUnimplementedMessage() {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		const messageService = <MessageService>(
			clientSession.activateService(MessageService.serviceName)!
		);
		await messageService.sendMessage(new TestMessage());

		// Wait for open channel response, to ensure the message is processed.
		await clientSession.openChannel();
	}

	@test
	public async sendDebugMessage() {
		const [clientSession, serverSession] = await this.createSessions();
		await connectSessionPair(clientSession, serverSession);

		const messageService = <MessageService>(
			clientSession.activateService(MessageService.serviceName)!
		);
		await messageService.sendMessage(new DebugMessage('test'));

		// Wait for open channel response, to ensure the message is processed.
		await clientSession.openChannel();
	}
}

class TestServiceConfig {}

@serviceActivation({ serviceRequest: TestService1.serviceName })
class TestService1 extends SshService {
	public static readonly serviceName = 'test-service-1';
	private readonly disposedCompletion = new PromiseCompletionSource<void>();

	public constructor(session: SshSession, config?: TestServiceConfig) {
		super(session);
		this.config = config;
	}

	public readonly config: TestServiceConfig | undefined;

	public get disposedPromise(): Promise<void> {
		return this.disposedCompletion.promise;
	}

	public isDisposed: boolean = false;

	public dispose(): void {
		this.isDisposed = true;
		this.disposedCompletion.resolve();
		super.dispose();
	}
}

@serviceActivation({ sessionRequest: TestService2.requestName })
class TestService2 extends SshService {
	public static readonly requestName = 'test-service-2';

	public constructor(session: SshSession, config?: TestServiceConfig) {
		super(session);
		this.config = config;
	}

	public readonly config: TestServiceConfig | undefined;

	public requestMessage: SessionRequestMessage | undefined;

	public async onSessionRequest(
		request: SshRequestEventArgs<SessionRequestMessage>,
	): Promise<void> {
		this.requestMessage = request.request;
		request.isAuthorized = true;

		if (request.request.wantReply) {
			await this.sendMessage(new SessionRequestSuccessMessage());
		}
	}
}

@serviceActivation({ channelType: TestService3.channelType })
class TestService3 extends SshService {
	public static readonly channelType = 'test-service-3';

	public constructor(session: SshSession, config?: TestServiceConfig) {
		super(session);
		this.config = config;
	}

	public readonly config: TestServiceConfig | undefined;

	public channel: SshChannel | undefined;

	public async onChannelOpening(args: SshChannelOpeningEventArgs): Promise<void> {
		this.channel = args.channel;
	}
}

@serviceActivation({ channelRequest: TestService4.requestName })
class TestService4 extends SshService {
	public static readonly requestName = 'test-service-4';

	public constructor(session: SshSession) {
		super(session);
	}

	public channel: SshChannel | undefined;

	public requestMessage: ChannelRequestMessage | undefined;

	public async onChannelRequest(
		channel: SshChannel,
		request: SshRequestEventArgs<ChannelRequestMessage>,
	): Promise<void> {
		this.channel = channel;
		this.requestMessage = request.request;
		request.isAuthorized = true;
	}
}

@serviceActivation({
	channelType: TestService5.channelType,
	channelRequest: TestService5.requestName,
})
class TestService5 extends SshService {
	public static readonly channelType = 'test-service-5-channel';
	public static readonly requestName = 'test-service-5';

	public constructor(session: SshSession) {
		super(session);
	}

	public channel: SshChannel | undefined;

	public requestMessage: ChannelRequestMessage | undefined;

	public async onChannelRequest(
		channel: SshChannel,
		request: SshRequestEventArgs<ChannelRequestMessage>,
	): Promise<void> {
		this.channel = channel;
		this.requestMessage = request.request;
		request.isAuthorized = true;
	}
}

@serviceActivation({ serviceRequest: MessageService.serviceName })
class MessageService extends SshService {
	public static readonly serviceName = 'test-message-service';

	public constructor(session: SshSession) {
		super(session);
	}

	public sendMessage(message: SshMessage, cancellation?: CancellationToken): Promise<void> {
		return super.sendMessage(message, cancellation);
	}
}

class TestMessage extends SshMessage {
	public get messageType(): number {
		return 199;
	}
	protected onRead(reader: SshDataReader): void {}

	protected onWrite(writer: SshDataWriter): void {}
}
